extensions/vscode/src/commands/stop-dev-server.ts

```diff
@@ -0,0 +1,190 @@
+        return;
+      case "Cancel":
+      default:
+        return;
+    }
+  }
+
+  const application =
+    applications.length === 1
+      ? applications[0]
+      : await quickPickApplication(applications);
+  if (!application) {
+    return;
+  }
+
+  const stopRequest = new StopDaemonRequest(
+    daemon.requestIdentifierGenerator.generate(),
+    application.id!
+  );
+
+  const deregisterPromise = onApplicationDeregistered(
+    daemon.applicationRegistry,
+    application
+  );
+
+  await window.withProgress(
+    {
+      location: 15,
+    },
+    async (progress) => {
+      progress.report({
+        message: `Stopping server...`,
+      });
+
+      const stopDaemonResponse = await daemon.send(stopRequest);
+      if (stopDaemonResponse.error) {
+        window.showErrorMessage(stopDaemonResponse.error.message);
+        return;
+      }
+
+      progress.report({
+        message: `Deregistering server...`,
+        increment: 75,
+      });
+
+      await deregisterPromise;
+
+      progress.report({
+        message: `Server stopped successfully`,
+        increment: 100,
+      });
+
+      // Add a small delay to allow the user to read the message.
+      return await new Promise((resolve) => setTimeout(resolve, 250));
+    }
+  );
+};
+
+class PickableDartFrogApplication implements QuickPickItem {
+  constructor(dartFrogApplication: DartFrogApplication) {
+    const addressWithoutProtocol = dartFrogApplication.address!.replace(
+      /.*?:\/\//g,
+      ""
+    );
+    this.label = `$(globe) ${addressWithoutProtocol}`;
+    this.description = dartFrogApplication.id?.toString();
+    this.application = dartFrogApplication;
+  }
+
+  public readonly application: DartFrogApplication;
+
+  label: string;
+  kind?: QuickPickItemKind | undefined;
+  description?: string | undefined;
+  detail?: string | undefined;
+  picked?: boolean | undefined;
+  alwaysShow?: boolean | undefined;
+  buttons?: readonly QuickInputButton[] | undefined;
+}
+
+/**
+ * Prompts the user to select a {@link DartFrogApplication} from a list of
+ * running {@link DartFrogApplication}s.
+ *
+ * @param applications The running {@link DartFrogApplication}s to choose from.
+ * @returns The selected {@link DartFrogApplication} or `undefined` if the user
+ * cancelled the selection.
+ */
+async function quickPickApplication(
+  applications: DartFrogApplication[]
+): Promise<DartFrogApplication | undefined> {
+  const quickPick = window.createQuickPick<PickableDartFrogApplication>();
+  quickPick.placeholder = "Select a server to stop";
+  quickPick.busy = true;
+  quickPick.ignoreFocusOut = true;
+  quickPick.items = applications.map(
+    (application) => new PickableDartFrogApplication(application)
+  );
+  quickPick.show();
+
+  return new Promise<DartFrogApplication | undefined>((resolve) => {
+    quickPick.onDidChangeSelection((value) => {
+      quickPick.dispose();
+
+      if (value.length === 0) {
+        resolve(undefined);
```

Comment:
```suggestion
        resolve();
```

`undefined` is already implicity

Reply:
This can't be done since the resolve function expects one argument.

Reply:
hmmm, got ya, TypeScript must define the types that way then, which makes no sense to me, since you can have a void Promise, this argument should be optional 🤔